compiler/rustc_lint/src/elided.rs

```diff
@@ -0,0 +1,118 @@
+use rustc_hir::def_id::LocalDefId;
+use rustc_hir::intravisit::{walk_fn_decl, FnKind, Visitor};
+use rustc_hir::{
+    Body, FnDecl, FnSig, ImplItem, ImplItemKind, Lifetime, LifetimeName, TraitItem, TraitItemKind,
+};
+use rustc_middle::ty::layout::HasTyCtxt;
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::symbol::kw;
+use rustc_span::Span;
+
+use crate::lints::{ElidedNamedLifetime, ElidedNamedLifetimeSuggestion};
+use crate::{LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `elided_named_lifetimes` lint detects when an elided
+    /// lifetime ends up being a named lifetime, such as `'static`
+    /// or some lifetime parameter `'a`.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(elided_named_lifetimes)]
+    /// struct Foo;
+    /// impl Foo {
+    ///     pub fn get_mut(&'static self, x: &mut u8) -> &mut u8 {
+    ///         unsafe { &mut *(x as *mut _) }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Lifetime elision is quite useful, because it frees you from having
+    /// to give each lifetime its own name, but sometimes it can produce
+    /// somewhat surprising resolutions. In safe code, it is mostly okay,
+    /// because the borrow checker prevents any unsoundness, so the worst
+    /// case scenario is you get a confusing error message in some other place.
+    /// But with `unsafe` code, such unexpected resolutions may lead to unsound code.
+    pub ELIDED_NAMED_LIFETIMES,
+    Warn,
+    "detects when an elided lifetime gets resolved to be `'static` or some named parameter"
+}
+
+declare_lint_pass!(ElidedNamedLifetimes => [ELIDED_NAMED_LIFETIMES]);
+
+impl<'tcx> LateLintPass<'tcx> for ElidedNamedLifetimes {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        _: LocalDefId,
+    ) {
+        self.check_decl(cx, decl)
+    }
+
+    fn check_trait_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx TraitItem<'tcx>) {
+        if let TraitItemKind::Fn(FnSig { decl, .. }, _) = item.kind {
+            self.check_decl(cx, decl)
+        }
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx ImplItem<'tcx>) {
+        if let ImplItemKind::Fn(FnSig { decl, .. }, _) = item.kind {
+            self.check_decl(cx, decl)
+        }
+    }
+}
+
+impl ElidedNamedLifetimes {
+    fn check_decl<'tcx>(&self, cx: &LateContext<'tcx>, decl: &'tcx FnDecl<'tcx>) {
+        walk_fn_decl(&mut LifetimeVisitor { cx }, decl)
+    }
+}
+struct LifetimeVisitor<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+}
+
+impl<'tcx> Visitor<'tcx> for LifetimeVisitor<'_, 'tcx> {
+    fn visit_lifetime(&mut self, lifetime: &'tcx Lifetime) -> Self::Result {
```

Comment:
Should we add a `check_lifetime` method to `LateLintPass`, and directly implement it? See the `late_lint_methods` macro.

Reply:
Heh, I was about to push just that